wallet-new/src/Cardano/Wallet/WalletLayer/ExecutionTimeLimit.hs

```diff
@@ -0,0 +1,36 @@
+module Cardano.Wallet.WalletLayer.ExecutionTimeLimit (
+      limitExecutionTimeTo
+    , TimeExecutionLimit(..)
+    ) where
+
+import           Universum
+
+import           Formatting (bprint, shown, (%))
+import qualified Formatting.Buildable
+
+import           Control.Concurrent (threadDelay)
+import           Control.Concurrent.Async (AsyncCancelled (..))
+import qualified Control.Concurrent.Async as Async
+import           Data.Time.Units (Second)
+
+data TimeExecutionLimit =
+    TimeExecutionLimitReached Second
+    deriving (Eq, Show)
+
+instance Exception TimeExecutionLimit
+
+instance Buildable TimeExecutionLimit where
+    build (TimeExecutionLimitReached secs) =
+        bprint ("TimeExecutionLimitReached " % shown) secs
+
+limitExecutionTimeTo :: Second
+                     -> (TimeExecutionLimit -> b)
+                     -> IO (Either b a)
+                     -> IO (Either b a)
+limitExecutionTimeTo secs onTimeLimit action = do
```

Comment:
As-it-is this function is quite subtle (and broken) in case pure values would be wrapped into an `IO` action, as those would remain as thunk with the risk of never being evaluated. At the bare minimum we want to at least run `evaluate` on the result of the input `action`.

Reply:
so...? :)